### DISCUSSION TITLE
Restrictions on `FullyConnectedTensorProduct` weights for maintaining equivariance

### DISCUSSION THREAD

**Discussion by ipcamit** [e3nn/e3nn#347] [Q&A] [answered]
I am trying to understand the limitations and restrictions on equivariant models for my use. For the same I tried to construct a  graph layer with following message ($r$ = radial basis function)
$\mathbf{m_{ij}} = \mathbf{h_i} \otimes\phi_e (\mathbf{h_j},r(\vec{r_{ij}})) Y^l_m (\hat{r_{ij}})$
Which is kind of attempted inspiration from [EGNN](https://arxiv.org/pdf/2102.09844.pdf) $\mathbf{m_{ij}} = \phi_e(\mathbf{h_i}^l, \mathbf{h_j}^i,\|r_i - r_j \|^2, a_{ij})$.

When I test its equivariance error, it shows that my layer is not equivariant. The culprit is $\phi_e(\mathbf{h_j},r(\vec{r_{ij}}))$ and I only get equivariance if my `FullyConnectedTensorProduct` weights are obtained from radial vectors, i.e $\phi_e(r(\vec{r_{ij}}))$. 

So I would like to understand what kind of restrictions are needed on tensor product weights to ensure equivariance? 

(I am not sure if it related to E3NN or general mathematics, so I would be happy even if you could point to appropriate resources)

**Comment by blondegeek** [comment]
In general, weights in the networks must be scalars. (There are very
special cases where one might also want to learn non-scalar weights but
this is typically not the case as it does break E(3) equivariance).

Specifically for the tensor product, the weights applied are to entire
"paths". For example, when you tensor product (FullTensorProduct in e3nn
language) two vectors (1o x 1o), you get scalar (0e), pseudovector (1e) and
traceless symmetric (2e) components. Each of these outputs is a path (1o x
1o -> 0e, 1o x 1o -> 1e, and 1o x 1o -> 2e). You can apply scalar weights
to each path and preserve equivariance. In the FullyConnectedTensorProduct
case, you can have many more or fewer paths than those in a conventional
tensor product and thus more scalar weights for each path.

It looks like you are trying to make the weights depend on both the radial
information and the feature at point j. In principle there should be
nothing wrong with this, however, the resulting information from \phi_e
does need to be a scalar and it needs to be applied to the correct pairwise
interaction. Without knowing more about \phi_e, it would be difficult for
me to say what specifically is causing the break in equivariance. If the
h_j is always in the same coordinate frame as the rest of the input, there
should be no issue with using it for the weights.

One way to think about weights needing to be scalars is the following: let
f be your layer function f: X, W -> Y, where x is your input, w your
weights, and y your output. To be equivariant to group G, f(D_X(g)x,
D_W(g)w) = D_Y(g)f(x, w) for group representations D_X, D_W, and D_Y and
for all g in G. The easiest way to guarantee this is that D_W = identity
and that W is a vector space of scalars under the action of G. Again,
non-scalar tensor weights are possible but that implies your network will
not be E(3) equivariant -- it will be equivariant under whatever symmetry
is defined by your tensor weights w, which is not necessary or desirable
for the vast majority of use cases.

Is that helpful?

**Reply by ipcamit** [reply]
Thank you for the detailed reply. So if I understand it correctly `FullyConnectedTensorProduct` by default uses scalar weights only and as a user I need not worry about it right? (I tried outputting it with fixed weights and I can see that it has one scalar per representation per path as weight). 
In that case my approach should be feasible. Input irreps are random features right now, but would represent atom embeddings later on.

Below is simplified class that tried to implement the above equivariant layer if that helps
```python
class E3NNGCL_Simple(gnn.MessagePassing):
    def __init__(self, irrep_in, irrep_out, irrep_Y):
        super().__init__(aggr="add")
        self.irrep_Y = irrep_Y
        self.radial_basis_length = 25

        self.otimes = o3.FullyConnectedTensorProduct(irreps_in1=irrep_in, 
                                                    irreps_in2=irrep_Y, 
                                                    irreps_out=irrep_out, 
                                                    shared_weights=False)#Tensor Product
        
        self.phi_e = enn.FullyConnectedNet([irrep_in.dim + self.radial_basis_length, 64,self.otimes.weight_numel], torch.relu) # Err ~ 7.0
        
        # self.phi_e = enn.FullyConnectedNet([self.radial_basis_length, 64, self.otimes.weight_numel], torch.relu) # Err ~ 1E-6
        
    def forward(self, h, pos, edge_index):
        h = self.propagate(h=h, pos=pos, edge_index=edge_index)
        return h
    
    def message(self, h_i, h_j, pos_i, pos_j):
        rij = pos_i - pos_j
        norm_rij = rij.pow(2).sum(dim=1).sqrt().unsqueeze(1)

        r_basis = gen_basis(norm_rij)
        rij_hat = rij/norm_rij
        h_j = self.phi_e(torch.cat([r_basis,h_j],1))
        # h_j = self.phi_e(r_basis) # Err ~ 1E-6

        Y = o3.spherical_harmonics(self.irrep_Y, rij_hat, normalize=True, normalization="norm")
        m_ij = self.otimes(h_i,Y,h_j)
        return m_ij

irreps_in = o3.Irreps("48x0e+16x1o")
irreps_Y = o3.Irreps.spherical_harmonics(lmax=1)
irreps_out = o3.Irreps("48x0e+16x1o")
```
I was trying to rationalize the error as the fact that because may be `h_j` is random feature and is passed from simple MLP, it does not retain equivariance, but `r_basis`, being obtained form well behaved `1x1o` `r_ij` feature, maintains it when passed through MLP. 

But if understood above answer correctly then it should not matter as ultimately scalar weights are being applied.

**Comment by blondegeek** [comment, marked as answer]
In the code you sent you are applying a fully connected network on features
that are not scalars so it will not be equivariant. Specifically here...

phi_e defined....
>> self.phi_e = enn.FullyConnectedNet([irrep_in.dim +
self.radial_basis_length, 64,self.otimes.weight_numel], torch.relu)
...and used...
>> h_j = self.phi_e(torch.cat([r_basis,h_j],1))

In e3nn, we only use traditional fully connected networks on scalar inputs
e.g. radial basis functions). We use e3nn.o3.Linear layers for general
equivariant linear layers, which can only mix channels within the same
irrep.

**Reply by ipcamit** [reply]
Ah! I got it now! 
Thank you for the clarification. During actual production run, `h_i` and `h_j` would be collection of scalars. But for equivariance testing I took higher features just to be sure. Will look in o3.Linear too.
Thanks you

**Comment by blondegeek** [comment]
I would say instead that the takeaway is that \phi_e needs to be an
equivariant function for the inputs you give it -- currently it is not
when non-scalars are given as input because you're using a traditional
fully connected net which is not equivariant for non-scalars. \phi_e needs
to satisfy the condition that \phi_e: X-> Y, \phi_e(D_X(g)x) =
D_Y(g)f\phi_ex) where D_X and D_Y are the representations of the Euclidean
group on the vector spaces you have chosen for the inputs and outputs of
\phi_e.
